File: 2024/08/test_day_08.py
import unittest
from typing import List

from day_08 import Solution

AOC_SIMPLE_INPUT_1 = """..........
...#......
..........
....a.....
..........
.....a....
..........
......#...
..........
.........."""
AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS = {'a': [(3, 4), (5, 5)]}
AOC_SIMPLE_INPUT_1_ANTINODE_LOCATIONS = [(1,3), (7,6)]

AOC_SIMPLE_INPUT_2 = """..........
...#......
#.........
....a.....
........a.
.....a....
..#.......
......#...
..........
.........."""
AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS = {'a': [(3, 4), (5, 5), (4, 8)]}
AOC_SIMPLE_INPUT_2_ANTINODE_LOCATIONS = [(1,3), (7,6), (2,0), (6,2)]

AOC_INPUT = """......#....#
...#....0...
....#0....#.
..#....0....
....0....#..
.#....A.....
...#........
#......#....
........A...
.........A..
..........#.
..........#."""
AOC_INPUT_ANTENNA_LOCATIONS = {'0':[(1,8), (2,5), (3,7), (4,4)],'A': [(5, 6), (8, 8), (9, 9)]}
AOC_INPUT_ANTINODE_LOCATIONS = [(0,6), (0,11), (1,3), (2,4), (2,10), (3,2), (4,9), (5,1), (5,6), (6,3), (7,0), (7,7), (10,10), (11,10)]

AOC_SIMPLE_RESONANT_INPUT = """T....#....
...T......
.T....#...
.........#
..#.......
..........
...#......
..........
....#.....
.........."""
AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS = {'T': [(0, 0),  (1, 3), (2, 1)]}
AOC_SIMPLE_RESONANT_INPUT_ANTINODE_LOCATIONS = [(0, 0), (0, 5), (1, 3), (2, 1), (2,6),
                                                     (3,9), (4,2), (6,3), (8,4)]

class TestSolution(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def test_load_map_simple(self):
        self.solution.load_map(AOC_SIMPLE_INPUT_1)
        self.assertEqual(self.solution.antennas.keys(), AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS.keys())
        for key in AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS.keys():
            self.assertCountEqual(self.solution.antennas[key], AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS[key])

    def test_find_antinodes_simple_1(self):
        self.solution.load_map(AOC_SIMPLE_INPUT_1)
        self.assertCountEqual(self.solution.antennas.keys(), AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS.keys())
        for key in AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS.keys():
            self.assertCountEqual(self.solution.antennas[key], AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS[key])
        self.assertEqual(len(self.solution.antinodes), len(AOC_SIMPLE_INPUT_1_ANTINODE_LOCATIONS))
        self.assertCountEqual(self.solution.antinodes, list(AOC_SIMPLE_INPUT_1_ANTINODE_LOCATIONS))

    def test_find_antinodes_simple_2(self):
        self.solution.load_map(AOC_SIMPLE_INPUT_2)
        self.assertCountEqual(self.solution.antennas.keys(), AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS.keys())
        for key in AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS.keys():
            self.assertCountEqual(self.solution.antennas[key], AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS[key])
        self.assertEqual(len(self.solution.antinodes), len(AOC_SIMPLE_INPUT_2_ANTINODE_LOCATIONS))
        self.assertCountEqual(self.solution.antinodes, AOC_SIMPLE_INPUT_2_ANTINODE_LOCATIONS)

    def test_find_antinodes_aoc_input(self):
        self.solution.load_map(AOC_INPUT)
        self.assertCountEqual(self.solution.antennas.keys(), AOC_INPUT_ANTENNA_LOCATIONS.keys())
        for key in AOC_INPUT_ANTENNA_LOCATIONS.keys():
            self.assertCountEqual(self.solution.antennas[key], AOC_INPUT_ANTENNA_LOCATIONS[key])
        self.assertEqual(len(self.solution.antinodes), len(AOC_INPUT_ANTINODE_LOCATIONS))
        self.assertCountEqual(self.solution.antinodes, AOC_INPUT_ANTINODE_LOCATIONS)

    def test_find_resonant_antinodes_simple_aoc_input(self):
        self.solution.load_map(AOC_SIMPLE_RESONANT_INPUT)
        self.assertCountEqual(self.solution.antennas.keys(), AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS.keys())
        for key in AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS.keys():
             self.assertCountEqual(self.solution.antennas[key], AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS[key])
        self.assertEqual(len(self.solution.resonant_antinodes), len(AOC_SIMPLE_RESONANT_INPUT_ANTINODE_LOCATIONS))
        self.assertCountEqual(self.solution.resonant_antinodes, AOC_SIMPLE_RESONANT_INPUT_ANTINODE_LOCATIONS)

if __name__ == '__main__':
    unittest.main()

File: 2024/08/test_day_08.jl
using Test
include("./day_08.jl")

AOC_SIMPLE_INPUT_1 = """..........
...#......
..........
....a.....
..........
.....a....
..........
......#...
..........
.........."""
AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS = Dict('a' => [(4, 5), (6, 6)])
AOC_SIMPLE_INPUT_1_ANTINODE_LOCATIONS = Set([(2,4), (8,7)])

AOC_SIMPLE_INPUT_2 = """..........
...#......
#.........
....a.....
........a.
.....a....
..#.......
......#...
..........
.........."""
AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS = Dict('a' => [(4, 5), (6, 6), (5, 9)])
AOC_SIMPLE_INPUT_2_ANTINODE_LOCATIONS = Set([(2,4), (8,7), (3,1), (7,3)])

AOC_INPUT = """......#....#
...#....0...
....#0....#.
..#....0....
....0....#..
.#....A.....
...#........
#......#....
........A...
.........A..
..........#.
..........#."""
AOC_INPUT_ANTENNA_LOCATIONS = Dict('0' => [(2,9), (3,6), (4,8), (5,5)],'A' => [(6, 7), (9, 9), (10, 10)])
AOC_INPUT_ANTINODE_LOCATIONS = Set([(1,7), (1,12), (2,4), (3,5), (3,11), (4,3), (5,10), (6,2), (6,7), (7,4), (8,1), (8,8), (11,11), (12,11)])

AOC_SIMPLE_RESONANT_INPUT = """T....#....
...T......
.T....#...
.........#
..#.......
..........
...#......
..........
....#.....
.........."""
AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS = Dict('T' => [(1, 1), (2, 4), (3, 2)])
AOC_SIMPLE_RESONANT_INPUT_ANTINODE_LOCATIONS = Set([(1, 1), (1, 6), (2, 4), (3, 2), (3,7),
                                                     (4,10), (5,3), (7,4), (9,5)])


function test_solution()

    @testset "Simple Input 1" begin
        solution = parse_input(AOC_SIMPLE_INPUT_1)
        @test solution.width == 10
        @test solution.height == 10
        antennas = find_antenna_locations(solution)
        @test keys(antennas)==keys(AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS)
        for key in keys(AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS)
            @test sort(antennas[key])==sort(AOC_SIMPLE_INPUT_1_ANTENNA_LOCATIONS[key])
        end
        @test find_antinode_locations(solution)==AOC_SIMPLE_INPUT_1_ANTINODE_LOCATIONS
    end

    @testset "Simple Input 2" begin
        solution = parse_input(AOC_SIMPLE_INPUT_2)
        antennas = find_antenna_locations(solution)
        @test keys(antennas)==keys(AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS)
        for key in keys(AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS)
            @test sort(antennas[key])==sort(AOC_SIMPLE_INPUT_2_ANTENNA_LOCATIONS[key])
        end
        @test find_antinode_locations(solution)==AOC_SIMPLE_INPUT_2_ANTINODE_LOCATIONS
    end

    @testset "AOC Part 1 Input" begin
        solution = parse_input(AOC_INPUT)
        antennas = find_antenna_locations(solution)
        @test keys(antennas)==keys(AOC_INPUT_ANTENNA_LOCATIONS)
        for key in keys(AOC_INPUT_ANTENNA_LOCATIONS)
            @test sort(antennas[key])==sort(AOC_INPUT_ANTENNA_LOCATIONS[key])
        end
        @test find_antinode_locations(solution)==AOC_INPUT_ANTINODE_LOCATIONS
    end

    @testset "Simple Resonant Input" begin
        solution = parse_input(AOC_SIMPLE_RESONANT_INPUT)
        @test find_antenna_locations(solution)==AOC_SIMPLE_RESONANT_INPUT_ANTENNA_LOCATIONS
        @test find_resonant_antinode_locations(solution)==AOC_SIMPLE_RESONANT_INPUT_ANTINODE_LOCATIONS
    end

    # @testset 
    # test_input = """120: 10 12
    #     4392: 48 13 72"""
    # operators = find_operators([String(strip(x)) for x in split(test_input, "\n")])
    # @test [x.operators for x in operators]==[['*'], ['+', '*']]
    # @test sum([x.target for x in operators if x.operators != []])==4512

    # operators = find_operators([String(strip(x)) for x in split(AOC_INPUT, "\n")])
    # @test [x.operators for x in operators]==AOC_OPERATORS
    # @test sum([x.target for x in operators if x.operators != []])==AOC_TARGET

    # operators = find_operators_concat([String(strip(x)) for x in split(AOC_CONCATENATION_INPUT, "\n")])
    # @test [x.operators for x in operators]==AOC_CONCATENATION_OPERATORS
    # @test sum([x.target for x in operators if x.operators != []])==AOC_CONCATENATION_TARGET
end

test_solution()
